### ISSUE TITLE
LoadBalancer shouldn't route traffic to tasks in certain states

### ISSUE THREAD

**Issue by stevvooe** [docker/docker#24092]
There are a few conditions where we should not route traffic to a task:
1. If the task is in a `STARTING` state.
2. If the task has a desired state of `SHUTDOWN` and is still `RUNNING`.

This will ensure that tasks will not receive traffic before they are ready and after they may be shutting down.

Let me know further details are needed.

@mrjana @aluzzardi

**Comment by stelund** [comment]
I guess it's not part of this issue, but it would be great to limit traffic to only when an task is considered healthy. A database for example can take some time to load up and start to listen on the network port.

It doesn't look like swarm has any task state for healthy though.

**Comment by aluzzardi** [comment]
@stelund A task will not progress to `RUNNING` if it's not healthy. Also, if health check fails, it will transition to `FAILED`.

**Comment by stelund** [comment]
@aluzzardi I've tried this with HEAD of master branch and to me it looks like a task enters running state before first health check. But then I am not able to show that an failed health check turns into an failed state either.

**Comment by ChristianKniep** [comment]
Intuitivly I would think that the unhealthy container should not be part of the load-balancer as @stelund wrote...
```
➜ /usr/local/bin/docker-1.12 ps
CONTAINER ID        IMAGE                                        COMMAND                  CREATED             STATUS                     PORTS               NAMES
43858ce2d53e        qnib/httpcheck:docker-1.12   "/opt/qnib/supervisor"   3 minutes ago       Up 3 minutes (unhealthy)   80/tcp              httpcheck.2.d39pw4km8lf8tvjzdr4l8eltk
23dae647f89e        qnib/httpcheck:docker-1.12   "/opt/qnib/supervisor"   8 minutes ago       Up 7 minutes (healthy)     80/tcp              httpcheck.1.4szm55i1dfvuzxmp73stw45wq
➜ curl -s server1:8080|grep HTTP
<h1>This is served via HTTP from 6d4288163773!</h1>
➜ curl -s server1:8080|grep HTTP
<h1>This is served via HTTP from 43858ce2d53e!</h1>
➜ curl -s server1:8080|grep HTTP
<h1>This is served via HTTP from 23dae647f89e!</h1>
➜ curl -s server1:8080|grep HTTP
<h1>This is served via HTTP from 6d4288163773!</h1>
➜ curl -s server1:8080|grep HTTP
<h1>This is served via HTTP from 43858ce2d53e!</h1>
➜ 
```

**Comment by sswierczyna** [comment]
Is it possible to disable routing traffic to container when port is not opened yet?

I have scenario in my dockerized application:

0. all ports in container are closed
1. start some service inside container
2. wait until service is started and ready to fulfil reqests
3. enable listener on port eg 8080 (open TCP socket)

I would be happy when traffic will be routed to container after than port is opened.

**Comment by aluzzardi** [comment]
Assigning to @dongluochen - however can't actually assign since he's not a maintainer of docker/docker

**Comment by xiaods** [comment]
any update?

**Comment by xiaods** [comment]
@aluzzardi any udpate on it?

**Comment by ChristianKniep** [comment]
@xiaods AFAIK the issue here is that it can not be handled by SwarmKit alone, which is only 'in power' of updating and handling the container state. Thus it can do stuff like handling rolling updates of certain tasks.
The loadbalancer is part of libnetwork and the state has to be recognized there.  Not that I am 100% certain about the exact workflow, but it assume it is part of this:

**Comment by aluzzardi** [comment]
@xios @ChristianKniep Correct - that needs to be handled in libnetwork.

We have a fix and are currently fixing it - perhaps @dongluochen could share more details

**Comment by xiaods** [comment]
any update?

**Comment by dongluochen** [comment]
@xiaods We are running tests on the PR and vendoring of libnetwork. The feature will be available in Docker-1.13.

**Comment by xiaods** [comment]
@dongluochen got it.
